nemo/collections/nlp/models/language_modeling/megatron_ptune_t5_model.py

```diff
@@ -0,0 +1,494 @@
+    average_losses_across_data_parallel_group,
+    build_position_ids,
+)
+from nemo.collections.nlp.modules.common.prompt_encoder import PromptEncoder
+from nemo.utils import logging
+
+try:
+    from apex.transformer import tensor_parallel
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_APEX = False
+
+
+__all__ = ['MegatronT5PTuneModel']
+
+
+class MegatronT5PTuneModel(NLPModel):
+    """
+    Megatron T5 P-Tune
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer)
+        initialize_model_parallel_for_nemo(
+            world_size=trainer.world_size,
+            global_rank=trainer.global_rank,
+            local_rank=trainer.local_rank,
+            tensor_model_parallel_size=cfg.get('tensor_model_parallel_size', 1),
+            seed=cfg.get('seed', 1234),
+        )
+
+        # shared params for dataset and data loaders
+        # tokenizer needs to get initialized before the super.__init__()
+        # as dataloaders and datasets need it to process the data
+
+        self.megatron_amp_o2 = cfg.get('megatron_amp_O2', False)
+        # TODO: Fix this once apex patches FusedScaledMaskedSoftmax.
+        # This is a workaround for the fact that `masked_softmax_fusion` has issues with certain input sizes that may be present while finetuning.
+        t5_cfg = MegatronT5Model.restore_from(
+            self.register_artifact('language_model.nemo_file', cfg.language_model.get('nemo_file', None)),
+            trainer=trainer,
+            return_config=True,
+        )
+        OmegaConf.set_struct(t5_cfg, True)
+        with open_dict(t5_cfg):
+            t5_cfg.masked_softmax_fusion = False
+            t5_cfg.megatron_amp_O2 = self.megatron_amp_o2
+
+        self.model = MegatronT5Model.restore_from(
+            self.register_artifact('language_model.nemo_file', cfg.language_model.get('nemo_file', None)),
+            trainer=trainer,
+            override_config_path=t5_cfg,
+        )
+
+        # self.model = MegatronT5Model.restore_from(
+        #     self.register_artifact('language_model.nemo_file', cfg.language_model.get('nemo_file', None)),
+        #     trainer=trainer)
+
+        self.tokenizer = self.model.tokenizer
+
+        self.float_type = self.model.enc_dec_model.enc_dec_model.encoder.model.layers[0].dtype
+
+        if not cfg.use_lm_finetune:
+            self.model.freeze()
+
+        hidden_size = self.model.cfg.hidden_size
+
+        # register the file containing the labels into the artifacts to get stored in the '.nemo' file later
+        self.word_embeddings = self.model.enc_dec_model.encoder_embedding.word_embeddings
+        self.position_embeddings = self.model.enc_dec_model.encoder_embedding.position_embeddings
+
+        # self.vocab = self.tokenizer.tokenizer.get_vocab()
+
+        self.template = cfg.prompt_encoder.template
+
+        self.prompt_encoder = PromptEncoder(
```

Comment:
Should `embed_input` be pushed into `PromptEncoder`?
This was prompt tuning will only require implementing a new class.

Reply:
`embed_input` should be put here. As it depends on a lot of class members. e.g. `self.word_embeddings`, `self.pseudo_token_id`. It will create a strong coupling if we pass those objects to the `prompt encoder`.